allinbits/cosmos-cash-poc
# feat(poa): initial implementation of poa consensus

### Description

- Initial implementation of poa consensus algorithm
- Update README.md to define goals
- Update spec to reflect code

I'm sorry for this monster PR, I promise this is the only PR that will be this big :grimacing: 

### Tasks

- https://app.asana.com/0/1188221391270721/1189252855298209
- https://app.asana.com/0/1188221391270721/1190234426690937

### How to test 

1. Ensure the install is working correctly

```sh
make install
```

```
poa -h
poacli -h
```

2. Ensure all the tests pass

```sh
make test
```

3. Initialize a development chain

```sh
make init-dev
```

4. Create a new validator using the poa module

```sh
make create-validator
```

5. Ensure the validator was created correctly and accepted is set to `false`

```sh
make query-validator
```

6. Vote for the previously created validator

```sh
make vote-validator
```

7. Ensure the vote was correctly cast

```sh
make query-vote
```

8. Check the validator has been accepted, validator.accepted should now be set to `true`

```sh
make query-validator
```

9. On every start block new votes are calculated and the validators are updated in the store

10. On every end block the new validator set is pushed to Tendermint if validator.Accepted == `true`
